controllers/azurecluster_controller_test.go

```diff
@@ -58,6 +81,186 @@ var _ = Describe("AzureClusterReconciler", func() {
 	})
 })
 
+func TestAzureClusterReconcile(t *testing.T) {
+	g := NewWithT(t)
+	scheme, err := newScheme()
+	g.Expect(err).NotTo(HaveOccurred())
+
+	defaultCluster := getFakeCluster()
+	defaultAzureCluster := getFakeAzureCluster()
+
+	cases := map[string]struct {
+		objects []runtime.Object
+		fail    bool
+		err     string
+		event   string
+	}{
+		"should reconcile normally": {
+			objects: []runtime.Object{
+				defaultCluster,
+				defaultAzureCluster,
+			},
+		},
+		"should raise event if the azure cluster is not found": {
+			objects: []runtime.Object{
+				defaultCluster,
+			},
+			event: "AzureClusterObjectNotFound",
+		},
+		"should raise event if cluster is not found": {
+			objects: []runtime.Object{
+				getFakeAzureCluster(func(ac *infrav1.AzureCluster) {
+					ac.OwnerReferences = nil
+				}),
+				defaultCluster,
+			},
+			event: "OwnerRefNotSet",
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			client := fake.NewClientBuilder().
+				WithScheme(scheme).
+				WithRuntimeObjects(tc.objects...).
+				WithStatusSubresource(
+					&infrav1.AzureCluster{},
+				).
+				Build()
+
+			reconciler := &AzureClusterReconciler{
+				Client:   client,
+				Recorder: record.NewFakeRecorder(128),
+			}
+
+			_, err := reconciler.Reconcile(context.Background(), ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: "default",
+					Name:      "my-azure-cluster",
+				},
+			})
+			if tc.event != "" {
+				g.Expect(reconciler.Recorder.(*record.FakeRecorder).Events).To(Receive(ContainSubstring(tc.event)))
+			}
+			if tc.fail {
+				g.Expect(err).To(MatchError(tc.err))
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+		})
+	}
+}
+
+func TestAzureClusterReconcileNormal(t *testing.T) {
+	cases := map[string]TestClusterReconcileInput{
+		"should reconcile normally": {
+			createAzureClusterService: func(cs *scope.ClusterScope) (*azureClusterService, error) {
+				return getDefaultAzureClusterService(func(acs *azureClusterService) {
+					acs.skuCache = resourceskus.NewStaticCache([]armcompute.ResourceSKU{}, cs.Location())
+					acs.scope = cs
+					acs.Reconcile = func(ctx context.Context) error {
+						return nil
+					}
+					acs.Delete = func(ctx context.Context) error {
+						return nil
+					}
+					acs.Pause = func(ctx context.Context) error {
+						return nil
+					}
```

Comment:
nit: Setting these as nil in `getDefaultAzureClusterService()` would clean up the code a bit. If there are multiple azureClusterServices that use the same fields, it would be better to include them in `getDefaultAzureClusterService()` by default rather than setting them each time. If you need to modify the field, setting them in the `changes()` func would overwrite it like before.